proposals/0000-autolinking-local-modules.md

```diff
@@ -0,0 +1,83 @@
+---
+title: Autolinking local modules
+author:
+- Szymon Rybczak
+- Satyajit Sahoo
+date: 21-08-2023
+---
+
+# RFC0000: Autolinking local modules
+
+## Summary
+
+This RFC proposes a new feature to [React Native CLI](https://github.com/react-native-community/cli/) where we can create a folder containing local native modules and views and have them linked automatically. Proposed solution will simplify the space of creating local native modules and views and using them later in the app.
+
+## Motivation
+
+Currently, when we have local native modules, we have to create them under `android` and `ios` folders and write the code for registering those packages manually. This poses a few problems:
+
+1. The new modules become part of the `android` and `ios` source code—which means upgrading the project to a newer React Native version is harder as it also needs to consider those changes,
+2. The module boilerplate is written manually, that's the place where users can make mistakes and it takes a lot of time to write one,
+3. The code for registering the modules is additional boilerplate which is not required for third party packages (because they're automatically linked),
+4. The workflow for creating new architecture modules without creating packages is not streamlined,
+5. It's challenging to open-source the local module, as it requires creating a dedicated library setup that adds extra friction to the process.
```

Comment:
I think I may be missing something because I'm struggling to understand the scenarios this is supposed to address. I have some questions for your points here:

1.  How does having local native modules work around the fact that you need to deal with React Native upgrades? If your modules are written as Turbo/Native Modules, you still need to consider changes in React Native?
    1. If the generated modules don't have a `package.json`, what's the difference between this vs. putting the files directly in your project?
2. I can see the benefit of having the modules automatically generated, but isn't this already resolved by TurboModule generation?
3. On iOS/macOS, modules are statically linked when you define them. There is no registration boilerplate. On Android, we do need to register them manually. Is this what you're trying to solve?
4. Is this true? Doesn't codegen address this? If codegen is missing anything, I think it makes more sense to address those gaps instead. @cortinico?
5. I'm struggling to see how this is better. See comments on point 1.

Reply:
> Doesn't codegen address this? If codegen is missing anything, I think it makes more sense to address those gaps instead. @cortinico?

Practically codegen can consume local modules already. You'll have to define at the app level where they're placed and they will be consumed, so I don't see any significant gaps.

I would refrain from extending codegen to search also into `local_modules` as that will complicate the infra, and I would just re-use the infra that we already have (i.e. let the module pass through `node_modules` or via the app project).

Reply:
> How does having local native modules work around the fact that you need to deal with React Native upgrades? If your modules are written as Turbo/Native Modules, you still need to consider changes in React Native?

It's not about upgrading the modules, it's about upgrading the app. If your custom native code is outside of android/ios folders, you can just recreate android/ios folders of your app as you don't need to account for other changes related to your custom native code and carefully review the diff.

> I can see the benefit of having the modules automatically generated, but isn't this already resolved by TurboModule generation

Not sure what you mean here. Do you mean codegen? The point is about the initial boilerplate that we write manually before we start working on the module.

> Is this true? Doesn't codegen address this? If codegen is missing anything, I think it makes more sense to address those gaps instead.

It's more about polishing the workflow. In this case, having same setup as a library means we can have the same workflow and any improvements would apply in both cases.
https://github.com/reactwg/react-native-new-architecture/discussions/142#discussioncomment-6583129

Reply:
1. The main point that local modules are solving is when it comes to upgrades we have less custom code to take into account, because the most common scenario is that users put modules code under `ios/` or `android/` directories. 
2. What I described in this point is not role of codegen. What we're trying to achieve with CRNL is that it will contain templates for all kinds of local modules, so we'll automate process of creating modules from scratch. So user will be able to create module with just one command, not as right now there's need to create all files manually (that's where people make mistakes). 
3. Yes, exactly. With [dependencies](https://github.com/react-native-community/cli/blob/main/docs/dependencies.md) logic, that we'll reuse for autolinking we don't need to integrate with code under `android` directory at all.
4. So right now this is doable, but it isn't documented anywhere. (https://github.com/reactwg/react-native-new-architecture/discussions/142#discussioncomment-6583129) To do so you need specify path to local modules in `react-native.config.js` under `dependencies` field. And codegen will take into account these dependencies, because of this [logic](https://github.com/facebook/react-native/blob/97efa25afaf97da45ce4f09ccd61423e62317ed3/packages/react-native/scripts/codegen/generate-artifacts-executor.js#L223-L270). What we would like to do is to CLI look under `local_modules` directory and automatically under the hood link modules as `dependencies`. But the only thing that we need to implement is to codegen look also under `local_modules` directory, because if not we loose the whole point of autolinking modules (because then we would need to manually specify them under `dependencies` field in `react-native.config.js`). I think it wouldn't complicate logic behind the codegen. I don't quite understand how would it work with _"let the module pass through `node_modules` or via the app project"_ 🤔

Reply:
> but it's not to clear to me whether you mean the user's app or the library's example app

the local modules are for using in the user's app. there won't be a separate example app here as the user would be using the library in their app. the modifications to CRNL would consist of removing the example app and any other library specific stuff.

Reply:
> … why is local modules better than just putting your modules directly in the project? I think you can still streamline this (e.g. with code generation) without introducing the concept of local modules.

What about this? I think if CRNL can just output those files directly into your project, you've already solved a pain point. I don't see why the concept of local modules is necessary.